### PR TITLE
implement FSDPTransform.transform_state_dict_for_submodule

### PR DESCRIPTION
**Pull request by t-vi**
This implements state dict transformation, immediately enabling load_original_state_dict to load an unsharded dictionary into a sharded model.

This is a step in #987 .

(it's ready but relative to #988)

### CODE REVIEW

File: thunder/distributed/__init__.py [hunk 1]
```diff
@@ -594,7 +594,29 @@ def _shard_param(
         # NOTE This could be a ShardTensor to indicate other parts of the code
         # that it's sharded and should be treated differently
         shard = param.data.narrow(dim_to_shard, chunk_size * rank, chunk_size).clone()
-        param.data = shard
+        return shard, None
+
+
+def _shard_param(
```
Comment (crcrpar) [inline thunder/distributed/__init__.py:600]: shouldn't this be decorated with `torch.no_grad()`?

File: thunder/distributed/__init__.py [hunk 2]
```diff
@@ -594,7 +594,29 @@ def _shard_param(
...
+    shard, padding_size = _shard_tensor(
+        param,
+        rank,
+        world_size,
+        name,
+        allow_padding_for_fsdp=allow_padding_for_fsdp,
+        dim=dim,
+    )
+    param.data = shard
+    if allow_padding_for_fsdp:
```
Comment (crcrpar) [inline thunder/distributed/__init__.py:618]: nit-picking
```suggestion
    if allow_padding_for_fsdp is not None:
```

File: thunder/distributed/transforms/fsdp_v2.py [hunk 3]
```diff
@@ -234,7 +243,14 @@ def transform_module(
     def transform_state_dict_for_submodule(
         self, model: thunder.ThunderModule, submodule_name: str, state_dict: dict
     ) -> dict:
-        raise NotImplementedError("cannot transform state dict yet")
+        prefix = submodule_name + ("." if submodule_name else "")
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:246]: nit-picking
```suggestion
        prefix = ""
        if submodule_name:
            prefix = f"{submodule_name}."
```

**Review by crcrpar — COMMENTED** [review]
so would this happen to indicate that all-gathering sharded state dict is user's responsibility?

File: thunder/distributed/__init__.py (hunk 2 above)
Comment (t-vi) [reply thunder/distributed/__init__.py:618]: This is a (non-optional) bool and defaults to false, I don't think it should be None.

File: thunder/distributed/__init__.py (hunk 1 above)
Comment (t-vi) [reply thunder/distributed/__init__.py:600]: Great idea!

File: thunder/distributed/__init__.py [hunk 4]
```diff
@@ -555,15 +556,15 @@ def _shard_params(
             sharded_params[param] = True
 
 
-def _shard_param(
+def _shard_tensor(
```
Comment (t-vi) [inline thunder/distributed/__init__.py:559]:
```suggestion
@torch.no_grad()
def _shard_tensor(
```

File: thunder/distributed/__init__.py (hunk 2 above)
Comment (crcrpar) [reply thunder/distributed/__init__.py:618]: my bad sorry I was confused with the other int | None padding var

File: thunder/distributed/__init__.py (hunk 1 above)
Comment (t-vi) [reply thunder/distributed/__init__.py:600]: I added the decorator to `_shard_tensor`.

**Review by crcrpar — APPROVED** [review]
LGTM

**Comment by t-vi** [comment]
Thank you @crcrpar
